# World Happiness Report - Exploratory Practice

**Dataset:** World Happiness Report (source: https://www.kaggle.com/datasets/unsdsn/world-happiness/data)

**Goal:** Practice pandas fundamentals (inspecting, filtering, sorting, grouping, merging) by exploring what factors relate to happiness scores across countries.

**Date started:** 2026-09-04

## 1. Imports & Setup

In [1]:
import pandas as pd 
import numpy as np

## 2. Loading Data

In [2]:
df_2019 = pd.read_csv('data/2019.csv')

## 3. Inspecting Data

**Q: What are the shape, column names, and dtypes of the dataframe?**

In [3]:
df_2019.shape

(156, 9)

In [4]:
df_2019.info()

<class 'pandas.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    str    
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     156 non-null    float64
dtypes: float64(7), int64(1), str(1)
memory usage: 11.1 KB


In [5]:
df_2019.head()

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption
0,1,Finland,7.769,1.340,1.587,0.986,0.596,0.153,0.393
1,2,Denmark,7.600,1.383,1.573,0.996,0.592,0.252,0.410
2,3,Norway,7.554,1.488,1.582,1.028,0.603,0.271,0.341
3,4,Iceland,7.494,1.380,1.624,1.026,0.591,0.354,0.118
4,5,Netherlands,7.488,1.396,1.522,0.999,0.557,0.322,0.298


In [6]:
df_2019.tail()

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption
151,152,Rwanda,3.334,0.359,0.711,0.614,0.555,0.217,0.411
152,153,Tanzania,3.231,0.476,0.885,0.499,0.417,0.276,0.147
153,154,Afghanistan,3.203,0.350,0.517,0.361,0.000,0.158,0.025
154,155,Central African Republic,3.083,0.026,0.000,0.105,0.225,0.235,0.035
155,156,South Sudan,2.853,0.306,0.575,0.295,0.010,0.202,0.091


**Dataset context:**
* There are 156 rows, each representing a country in the dataset, and 9 columns. 
* The first 3 columns are the country's rank, the country, and its happiness score.
* The next 6 columns describe the extent to which each of 6 factors - GDP, social support, life expectancy, freedom, generosity, and absence of corruption - contribute to making life evaluations higher in each country than they are in Dystopia. 
* Dystopia is a hypothetical country that has values equal to the world's lowest national averages for each of the 6 factors. It is a benchmark against which other countries are compared. 

**Q: Are there are any misisng values, and in which columns?**

In [7]:
df_2019.sample(10)

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption
36,37,Bahrain,6.199,1.362,1.368,0.871,0.536,0.255,0.110
140,141,Liberia,3.975,0.073,0.922,0.443,0.370,0.233,0.033
148,149,Syria,3.462,0.619,0.378,0.440,0.013,0.331,0.141
35,36,Italy,6.223,1.294,1.488,1.039,0.231,0.158,0.030
149,150,Malawi,3.410,0.191,0.560,0.495,0.443,0.218,0.089
44,45,Nicaragua,6.105,0.694,1.325,0.835,0.435,0.200,0.127
85,86,Kyrgyzstan,5.261,0.551,1.438,0.723,0.508,0.300,0.023
16,17,Germany,6.985,1.373,1.454,0.987,0.495,0.261,0.265
129,130,Sri Lanka,4.366,0.949,1.265,0.831,0.470,0.244,0.047
151,152,Rwanda,3.334,0.359,0.711,0.614,0.555,0.217,0.411


In [8]:
# checks if there are NaN values in any of the columns
df_2019.isna().any()

Overall rank                    False
Country or region               False
Score                           False
GDP per capita                  False
Social support                  False
Healthy life expectancy         False
Freedom to make life choices    False
Generosity                      False
Perceptions of corruption       False
dtype: bool

* `.isna()` returns a dataframe the same shape as `df_2019`, filled with `True`/`False`:
    * `True` wherever the original dataframe has a missing value (`NaN`)
    * `False` everywhere else

* `.any()` returns whether any element is `True` in a series or dataframe

There does not seem to be any missing values. The dataset looks well cleaned.

**Q: What is the mean, min, and max of the happiness score column?**

In [9]:
df_2019['Score'].describe()

count    156.000000
mean       5.407096
std        1.113120
min        2.853000
25%        4.544500
50%        5.379500
75%        6.184500
max        7.769000
Name: Score, dtype: float64

In [10]:
df_2019['Score'].mean()

np.float64(5.407096153846155)

In [11]:
df_2019['Score'].min()

np.float64(2.853)

In [12]:
df_2019['Score'].max()

np.float64(7.769)

* Mean happiness score: 5.407
* Min happiness score: 2.853
* Max happiness score: 7.769

## 4. Cleaning Data

In [13]:
df_2019

,Overall rank,Country or region,Score,GDP per capita,Social support,Healthy life expectancy,Freedom to make life choices,Generosity,Perceptions of corruption
0,1,Finland,7.769,1.340,1.587,0.986,0.596,0.153,0.393
1,2,Denmark,7.600,1.383,1.573,0.996,0.592,0.252,0.410
2,3,Norway,7.554,1.488,1.582,1.028,0.603,0.271,0.341
3,4,Iceland,7.494,1.380,1.624,1.026,0.591,0.354,0.118
4,5,Netherlands,7.488,1.396,1.522,0.999,0.557,0.322,0.298
...,...,...,...,...,...,...,...,...,...
151,152,Rwanda,3.334,0.359,0.711,0.614,0.555,0.217,0.411
152,153,Tanzania,3.231,0.476,0.885,0.499,0.417,0.276,0.147
153,154,Afghanistan,3.203,0.350,0.517,0.361,0.000,0.158,0.025
154,155,Central African Republic,3.083,0.026,0.000,0.105,0.225,0.235,0.035


Rename column headers to all lowercase and without spaces, so they can be more conveniently accessed later.

In [14]:
mapping = {
    'Overall rank' : 'overall_rank',
    'Country or region' : 'country',
    'Score' : 'score',
    'GDP per capita' : 'gdp_per_capita',
    'Social support' : 'social_support',
    'Healthy life expectancy' : 'life_expectancy',
    'Freedom to make life choices' : 'freedom',
    'Generosity' : 'generosity',
    'Perceptions of corruption' : 'perceived_corruption'
}

df_2019 = df_2019.rename(columns=mapping)

df_2019.head()

,overall_rank,country,score,gdp_per_capita,social_support,life_expectancy,freedom,generosity,perceived_corruption
0,1,Finland,7.769,1.340,1.587,0.986,0.596,0.153,0.393
1,2,Denmark,7.600,1.383,1.573,0.996,0.592,0.252,0.410
2,3,Norway,7.554,1.488,1.582,1.028,0.603,0.271,0.341
3,4,Iceland,7.494,1.380,1.624,1.026,0.591,0.354,0.118
4,5,Netherlands,7.488,1.396,1.522,0.999,0.557,0.322,0.298


## 5. Analysis: Sorting & Filtering

**Q: Create a dataframe of the 10 happiest countries, keeping only the rank, country name, and score columns**

In [15]:
# get top 10 ranked countries using boolean indexing
df_top_10_countries = df_2019[df_2019['overall_rank'] <= 10]

# alternatively: df_2019.head(10)

# keep only desired columns by indexing with list of column names
df_top_10_countries = df_top_10_countries[['overall_rank', 'country', 'score']]

# reset index to start from 0, and drop old index column
df_top_10_countries.reset_index(drop=True, inplace=True)

df_top_10_countries

,overall_rank,country,score
0,1,Finland,7.769
1,2,Denmark,7.600
2,3,Norway,7.554
3,4,Iceland,7.494
4,5,Netherlands,7.488
5,6,Switzerland,7.480
6,7,Sweden,7.343
7,8,New Zealand,7.307
8,9,Canada,7.278
9,10,Austria,7.246


**Q: Create a dataframe of the 10 least happy countries.**

In [16]:
# get lowest 10 ranked countries
df_lowest_10_countries = df_2019.tail(10)

# keep only desired columns by indexing with list of column names
df_lowest_10_countries = df_lowest_10_countries[['overall_rank', 'country', 'score']]

# reset index to start from 0, drop old index colum
df_lowest_10_countries.reset_index(drop=True, inplace=True)

df_lowest_10_countries

,overall_rank,country,score
0,147,Haiti,3.597
1,148,Botswana,3.488
2,149,Syria,3.462
3,150,Malawi,3.410
4,151,Yemen,3.380
5,152,Rwanda,3.334
6,153,Tanzania,3.231
7,154,Afghanistan,3.203
8,155,Central African Republic,3.083
9,156,South Sudan,2.853


**Q: Among countries with GDP per capita above the median, which 10 have the lowest happiness score?**

In [17]:
# calculate median GDP per capita
median_gdp = df_2019['gdp_per_capita'].median()

# get countries with GDP per capita above the median
df_gdp_above_median = df_2019[df_2019['gdp_per_capita'] > median_gdp]

# among those, get 10 countries with lowest happiness score
df_gdp_above_median.nsmallest(10, 'score')

# alternatively: df_gdp_above_median.tail(10)

,overall_rank,country,score,gdp_per_capita,social_support,life_expectancy,freedom,generosity,perceived_corruption
147,148,Botswana,3.488,1.041,1.145,0.538,0.455,0.025,0.100
125,126,Iraq,4.437,1.043,0.980,0.574,0.241,0.148,0.089
116,117,Iran,4.548,1.100,0.842,0.785,0.305,0.270,0.125
103,104,Gabon,4.799,1.057,1.183,0.571,0.295,0.043,0.055
96,97,Bulgaria,5.011,1.092,1.513,0.815,0.311,0.081,0.004
92,93,China,5.191,1.029,1.125,0.893,0.521,0.058,0.100
90,91,Lebanon,5.197,0.987,1.224,0.815,0.216,0.166,0.027
89,90,Azerbaijan,5.208,1.043,1.147,0.769,0.351,0.035,0.182
87,88,Algeria,5.211,1.002,1.160,0.785,0.086,0.073,0.114
86,87,Turkmenistan,5.247,1.052,1.538,0.657,0.394,0.244,0.028


`nsmallest()` returns the `n` rows with the smallest values from a dataframe or series
* Syntax: `df.nsmallest(n, columns, keep)`
* Parameters: 
    * `n`: number of items to retrieve
    * `columns`: column name, or list of column names
    * `keep`: specifies what to do with duplicate rows

## 6. Analysis: Grouping

**Q: Split countries into low, medium, and high "tiers" based on the GDP per capita score contribution (hint: using pd.cut or pd.qcut to create the buckets). What's the average happiness score in each tier?**

In [18]:
df_2019['gdp_per_capita'].describe()

count    156.000000
mean       0.905147
std        0.398389
min        0.000000
25%        0.602750
50%        0.960000
75%        1.232500
max        1.684000
Name: gdp_per_capita, dtype: float64

In [37]:
# create labels for tier categories
labels = ['Low', 'Medium', 'High']

# get a subset of df_2019 containing this list of columns
df_countries_by_gdp = df_2019[['overall_rank', 'country', 'score', 'gdp_per_capita']]

# divide gdp values into 3 equally sized tiers (low, medium, high)
# add a new column for these tiers
df_countries_by_gdp['gdp_tier'] = pd.qcut(df_2019['gdp_per_capita'], q=3, labels=labels)

df_countries_by_gdp

,overall_rank,country,score,gdp_per_capita,gdp_tier
0,1,Finland,7.769,1.340,High
1,2,Denmark,7.600,1.383,High
2,3,Norway,7.554,1.488,High
3,4,Iceland,7.494,1.380,High
4,5,Netherlands,7.488,1.396,High
...,...,...,...,...,...
151,152,Rwanda,3.334,0.359,Low
152,153,Tanzania,3.231,0.476,Low
153,154,Afghanistan,3.203,0.350,Low
154,155,Central African Republic,3.083,0.026,Low


In [ ]:
# group the countries by the tier values
# for each group (low, medium, high), calculate the mean happiness score
df_countries_by_gdp.groupby('gdp_tier')['score'].mean() # returns a Series

gdp_tier
Low       4.428077
Medium    5.319365
High      6.473846
Name: score, dtype: float64

In [45]:
# convert series returned from groupby and aggregation into a dataframe
df_avg_score_by_gdp = df_countries_by_gdp.groupby('gdp_tier')['score'].mean().reset_index()
df_avg_score_by_gdp

,gdp_tier,score
0,Low,4.428077
1,Medium,5.319365
2,High,6.473846


* Countries were split into 3 equally-sized tiers based on GDP per capita. This was done using the `pd.qcut()` function, which splits a Series into approximately equally sized bins.
* A dataframe with only the `overall_rank`, `country`, `score`, and `gdp_per_capita` columns was obtained. A new `gdp_tier` column was added to display the GDP tier of each country.
* The countries were grouped by their `gdp_tier` column values. For each tier group (low, medium, high), the average happiness score was calculated by applying the `.mean()` function on the `score` column.
* The Series returned from the `groupby()` and `mean()` operations was converted to a DataFrame by applying the `.reset_index()` function. The index labels (0, 1, 2...) were pulled into a new column, changing the data structure from 1D (Series) to 2D (DataFrame).

Result: 
* Mean score for low tier countries: 4.43
* Mean score for medium tier countries: 5.32
* Mean score for high tier countries: 6.47

**Q: Repeat the same steps for the other 5 factors. Store the results in their own DataFrames.**

In [63]:
labels = ['Low', 'Medium', 'High']
factors = ['gdp_per_capita', 'social_support', 'life_expectancy', 'freedom', 'generosity', 'perceived_corruption']
# created dictionary of dataframes
dfs_avg_score_by_tier = {}

# for each factor, dynamically create its dataframe of average score by tier
# store each created dataframe into a dictionary of dataframes
for factor in factors:
    # create dataframe
    dfs_avg_score_by_tier[f'df_avg_score_by_{factor}'] = df_2019.groupby(
        pd.qcut(df_2019[factor], q=3, labels=labels)
        )['score'].mean().reset_index()

# verify by accessing a dataframe in dataframes dictionary
dfs_avg_score_by_tier['df_avg_score_by_social_support'].head()

,social_support,score
0,Low,4.342596
1,Medium,5.441288
2,High,6.437404


* The goal was to create 6 DataFrames that show the average score by tier, one for each factor. To do so, I looped through each factor in a list of factors, and inserted the factor name where needed dynamically using f-strings. 
* The line that creates the DataFrame is complex-looking, but it's just the combined version of what was done previously for the GDP per capita, but with the added use of f-strings. 